### PR TITLE
[FRONTEND][TF] conv2d_transpose 'SAME' support kernel more than 1x1

### PR DESCRIPTION
**Pull request by optima2005**
This is to support 'SAME' padding for conv3d_transpose for kernel more than 1x1

Discussions:
[https://discuss.tvm.ai/t/why-we-only-support-kernel-1-1-for-tf-conv2d-transpose-same/4957](https://discuss.tvm.ai/t/why-we-only-support-kernel-1-1-for-tf-conv2d-transpose-same/4957)

@yongwww @apivovarov

### CODE REVIEW

File: tests/python/frontend/tensorflow/test_forward.py [hunk 1]
```diff
@@ -362,10 +362,16 @@ def test_forward_convolution():
...
                           'NCHW', [4, 176, 8, 8])
+        _test_convolution('conv_transpose', [4, 32, 8, 8], [2, 2, 176, 32], [1, 1], [1, 1], 'SAME',
+                          'NCHW', [4, 176, 8, 8])
+        _test_convolution('conv_transpose', [4, 32, 8, 8], [3, 3, 176, 32], [1, 1], [1, 1], 'SAME',
+                          'NCHW', [4, 176, 8, 8])
         _test_convolution('conv_transpose', [4, 19, 8, 8], [3, 3, 19, 19], [1, 1], [2, 2], 'VALID',
                           'NCHW', [4, 19, 17, 17])
         _test_convolution('conv_transpose', [4, 19, 17, 17], [1, 1, 124, 19], [1, 1], [1, 1], 'SAME',
                           'NCHW', [4, 124, 17, 17])
+        _test_convolution('conv_transpose', [4, 19, 17, 17], [3, 3, 124, 19], [1, 1], [1, 1], 'SAME',
```
Comment (apivovarov) [inline tests/python/frontend/tensorflow/test_forward.py]: Can you also add test case for 
```
kernel 2x2, strides 2x2, SAME
kernel 3x3, strides 2x2, SAME
```
E.g. if input is 5x5 then valid outputs are 9x9 or 10x10 (you can use one or another in the output_shape tensor) regardless of the kernel size.

File: src/relay/op/nn/convolution.cc [hunk 2]
```diff
@@ -248,10 +248,17 @@ bool Conv2DTransposeRel(const Array<Type>& types,
   }
   // dilation
   Array<IndexExpr> oshape({dshape_nchw[0], channels, 0, 0});
-  oshape.Set(2, (param->strides[0] * (dshape_nchw[2] - 1) + dilated_ksize_y -
-                 2 * param->padding[0] + param->output_padding[0]));
-  oshape.Set(3, (param->strides[1] * (dshape_nchw[3] - 1) + dilated_ksize_x -
-                 2 * param->padding[1] + param->output_padding[1]));
+  if ( param->padding.size() == 2 ) {
```
Comment (apivovarov) [inline src/relay/op/nn/convolution.cc]: minor formatting comment. I think we do not need spaces after `(` and before `)`

File: src/relay/op/nn/convolution.cc [hunk 3]
```diff
@@ -248,10 +248,17 @@ bool Conv2DTransposeRel(const Array<Type>& types,
...
+    oshape.Set(2, (param->strides[0] * (dshape_nchw[2] - 1) + dilated_ksize_y -
+                   2 * param->padding[0] + param->output_padding[0]));
+    oshape.Set(3, (param->strides[1] * (dshape_nchw[3] - 1) + dilated_ksize_x -
+                   2 * param->padding[1] + param->output_padding[1]));
+  } else if (param->padding.size() == 4) {
+    oshape.Set(2, (param->strides[0] * (dshape_nchw[2] - 1) + dilated_ksize_y -
+                   param->padding[0] - param->padding[2] + param->output_padding[0]));
+    oshape.Set(3, (param->strides[1] * (dshape_nchw[3] - 1) + dilated_ksize_x -
+                   param->padding[1] - param->padding[3] + param->output_padding[1]));
```
Comment (apivovarov) [inline src/relay/op/nn/convolution.cc]: Formulas are very similar except of  using 
```
2 * param->padding[0]  vs  param->padding[0] + param->padding[2]
and 
2 * param->padding[1]   vs   param->padding[1] + param->padding[3]
```
Can we calculate correct padding and then use the same formulas with calculated padding for both cases?

File: tests/python/frontend/tensorflow/test_forward.py (hunk 1 above)
Comment (optima2005) [reply tests/python/frontend/tensorflow/test_forward.py]: The strides in transpose convolution is used to dilate the input, see [this](https://github.com/apache/incubator-tvm/blob/203d2c922dd09298eedf43ae6b0224c60359b169/topi/python/topi/nn/conv2d_transpose.py#L64).  If the input is dilated, there would be no way to get 'SAME' size of output by only padding.

File: src/relay/op/nn/convolution.cc (hunk 2 above)
Comment (optima2005) [reply src/relay/op/nn/convolution.cc]: I will fix it.

File: src/relay/op/nn/convolution.cc (hunk 3 above)
Comment (optima2005) [reply src/relay/op/nn/convolution.cc]: The additional condition is to handle the head padding and tail padding are diffenct, for even kernel in this case. I didn't quite understand your point. Please clearify.

File: tests/python/frontend/tensorflow/test_forward.py (hunk 1 above)
Comment (optima2005) [reply tests/python/frontend/tensorflow/test_forward.py]: ok. I can get your point. 'SAME' means the size of the enlarged input size(by dilation).

**Comment by FrozenGene** [comment]
@optima2005 do you mind taking up this RFC? https://github.com/apache/incubator-tvm/issues/2682 Because I see you start to support 4D padding in Conv2dTranspose. I think the work should be similar with convolution, which could make us avoid `pad` operator and get potential performance.

**Comment by optima2005** [comment]
@FrozenGene let me do some evaluations first and then I will get back to you if I think I could do it. Actually I am still in the progess to understand those ops' implementation, especially those optimazation approach.

File: tests/python/frontend/tensorflow/test_forward.py (hunk 1 above)
Comment (apivovarov) [reply tests/python/frontend/tensorflow/test_forward.py]: TF code for kernel 2x2, strides 2x2, SAME
```
dshape=(1,5,5,4)
#hwoi
kshape=(2,2,2,4)
oshape=(1,9,9,2)
# or 
oshape=(1,10,10,2)
dtype='float32'
with tf.Session() as sess:
    x = tf.placeholder(shape=dshape, dtype=dtype)
    w = tf.placeholder(shape=kshape, dtype=dtype)
    dc = tf.nn.conv2d_transpose(x, w, output_shape=oshape, strides=(1,2,2,1), padding='SAME')

    res_dc = sess.run(dc, feed_dict={x: data, w:weight})
```

File: src/relay/op/nn/convolution.cc (hunk 3 above)
Comment (apivovarov) [reply src/relay/op/nn/convolution.cc]: What about the following?
```
  if (param->padding.size() == 2) {
    hpad = 2 * param->padding[0];
    wpad = 2 * param->padding[1];
  } else if (param->padding.size() == 4) {
    hpad = param->padding[0] + param->padding[2];
    wpad = param->padding[1] + param->padding[3];
  }
  oshape.Set(2, (param->strides[0] * (dshape_nchw[2] - 1) + dilated_ksize_y -
                 hpad + param->output_padding[0]));
  oshape.Set(3, (param->strides[1] * (dshape_nchw[3] - 1) + dilated_ksize_x -
                 wpad + param->output_padding[1]));
```

Comment (optima2005) [reply src/relay/op/nn/convolution.cc]: This is better.  Thanks! 
But the input padding tuple is (top, left, bottom, right).  So, 
```
  } else if (param->padding.size() == 4) {
    pad_h = param->padding[0] + param->padding[2];
    pad_w = param->padding[1] + param->padding[3];
  } else {
```

**Comment by optima2005** [comment]
@apivovarov I added the test cases per as your proposal and tried to pass them. But one condition (10x10 in your example) still not working in cuda target. I haven't found the root cause. I am wondering whether you could help.

**Comment by optima2005** [comment]
I got this error on cuda target with test case:
```
        _test_convolution('conv_transpose', [4, 32, 8, 8], [3, 3, 176, 32], [1, 1], [2, 2], 'SAME',
                          'NCHW', [4, 176, 16, 16])
```
incubator-tvm/src/pass/loop_partition.cc:544: Cannot prove: ((((floordiv(((((floordiv((floormod(dh, 2) + 7), 2)*16) + 16)*(floordiv((floormod(dw, 2) + 7), 2) + 1)) + 63), 64) - 1) - ((((floordiv((floormod(dh, 2) + 7), 2)*16) + 16)*(floordiv((floormod(dw, 2) + 7), 2) + 1)) - (floordiv(((((floordiv((floormod(dh, 2) + 7), 2)*16) + 16)*(floordiv((floormod(dw, 2) + 7), 2) + 1)) + 63), 64)*63))) + 1) >= 0), when generating the post doubt loop

Is it relative to issue #4470?  @apivovarov

**Comment by optima2005** [comment]
This can be confirmed that after I add exclusion for this specific case in workaround #4472, that test case can pass.

**Comment by tqchen** [comment]
please comment about the status of the PR, @FrozenGene

File: tests/python/frontend/tensorflow/test_forward.py [hunk 4]
```diff
@@ -388,10 +401,23 @@ def test_forward_convolution():
...
+    _test_convolution('conv_transpose', [4, 8, 8, 32], [2, 2, 176, 32], [1, 1], [1, 1], 'SAME',
+                      'NHWC', [4, 8, 8, 176])
+    _test_convolution('conv_transpose', [4, 8, 8, 32], [2, 2, 176, 32], [1, 1], [2, 2], 'SAME',
+                      'NHWC', [4, 15, 15, 176])
+    _test_convolution('conv_transpose', [4, 8, 8, 32], [3, 3, 176, 32], [1, 1], [1, 1], 'SAME',
+                      'NHWC', [4, 8, 8, 176])
+    _test_convolution('conv_transpose', [4, 8, 8, 32], [3, 3, 176, 32], [1, 1], [2, 2], 'SAME',
+                      'NHWC', [4, 15, 15, 176])
+    # cuda target not working
+    #_test_convolution('conv_transpose', [4, 8, 8, 32], [3, 3, 176, 32], [1, 1], [2, 2], 'SAME',
```
Comment (FrozenGene) [inline tests/python/frontend/tensorflow/test_forward.py]: How about is?

File: tests/python/frontend/tensorflow/test_forward.py [hunk 5]
```diff
@@ -362,10 +362,23 @@ def test_forward_convolution():
...
                           'NCHW', [4, 176, 8, 8])
+        _test_convolution('conv_transpose', [4, 32, 8, 8], [2, 2, 176, 32], [1, 1], [1, 1], 'SAME',
+                          'NCHW', [4, 176, 8, 8])
+        _test_convolution('conv_transpose', [4, 32, 8, 8], [2, 2, 176, 32], [1, 1], [2, 2], 'SAME',
+                          'NCHW', [4, 176, 15, 15])
+        _test_convolution('conv_transpose', [4, 32, 8, 8], [3, 3, 176, 32], [1, 1], [1, 1], 'SAME',
+                          'NCHW', [4, 176, 8, 8])
+        _test_convolution('conv_transpose', [4, 32, 8, 8], [3, 3, 176, 32], [1, 1], [2, 2], 'SAME',
+                          'NCHW', [4, 176, 15, 15])
+        # cuda target not working
```
Comment (yongwww) [inline tests/python/frontend/tensorflow/test_forward.py]: What is the reason why it doen't work for cuda

File: tests/python/frontend/tensorflow/test_forward.py (hunk 4 above)
Comment (optima2005) [reply tests/python/frontend/tensorflow/test_forward.py]: The case failed due to the "Cannot prove" error when compiling the module. see  #4470. It seems there is no fix for it yet.

File: tests/python/frontend/tensorflow/test_forward.py (hunk 5 above)
Comment (optima2005) [reply tests/python/frontend/tensorflow/test_forward.py]: The case failed due to the "Cannot prove" error when compiling the module. see #4470. It seems there is no fix for it yet.

**Comment by optima2005** [comment]
For stride '2x2', I tried kernel size up to '7x7', all has the "Cannot prove" compiling error for cuda target. So I added the walkaround. for other strides > '2x2',  I didn't find this error. 
@yongwww  @FrozenGene  @apivovarov  Would you please take a look again? Thanks!

**Comment by optima2005** [comment]
@yzhliu, would you mind to take a look? It is the fix to 'SAME' mode of #4300 
Thanks!

**Comment by tqchen** [comment]
tHANKS @optima2005 @yongwww @FrozenGene
